pyomo/gdp/plugins/multiple_bigm.py

```diff
@@ -0,0 +1,724 @@
+class MultipleBigMTransformation(Transformation):
+    """
+    Implements the multiple big-M transformation from [1]. Note that this
+    transformation is no different than the big-M transformation for two-
+    term disjunctions, but that it may provide a tighter relaxation for
+    models containing some disjunctions with three or more terms.
+
+    [1] Francisco Trespalaios and Ignacio E. Grossmann, "Improved Big-M
+        reformulation for generalized disjunctive programs," Computers and
+        Chemical Engineering, vol. 76, 2015, pp. 98-103.
+    """
+
+    CONFIG = ConfigBlock('gdp.mbigm')
+    CONFIG.declare('targets', ConfigValue(
+        default=None,
+        domain=target_list,
+        description="target or list of targets that will be relaxed",
+        doc="""
+
+        This specifies the list of components to relax. If None (default), the
+        entire model is transformed. Note that if the transformation is done out
+        of place, the list of targets should be attached to the model before it
+        is cloned, and the list will specify the targets on the cloned
+        instance."""
+    ))
+    CONFIG.declare('assume_fixed_vars_permanent', ConfigValue(
+        default=False,
+        domain=bool,
+        description="Boolean indicating whether or not to transform so that "
+        "the transformed model will still be valid when fixed Vars are "
+        "unfixed.",
+        doc="""
+        This is only relevant when the transformation will be calculating M
+        values. If True, the transformation will calculate M values assuming
+        that fixed variables will always be fixed to their current values. This
+        means that if a fixed variable is unfixed after transformation, the
+        transformed model is potentially no longer valid. By default, the
+        transformation will assume fixed variables could be unfixed in the
+        future and will use their bounds to calculate the M value rather than
+        their value. Note that this could make for a weaker LP relaxation
+        while the variables remain fixed.
+        """
+    ))
+    CONFIG.declare('solver', ConfigValue(
+        default=SolverFactory('gurobi'),
+        description="A solver to use to solve the continuous subproblems for "
+        "calculating the M values",
+    ))
+    CONFIG.declare('bigM', ConfigValue(
+        default=None,
+        domain=_to_dict,
+        description="Big-M values to use while relaxing constraints",
+        doc="""
+        A user-specified dict or ComponentMap mapping tuples of Constraints
+        and Disjuncts to Big-M values valid for relaxing the constraint if
+        the Disjunct is chosen.
+
+        Note: Unlike in the bigm transformation, we require the keys in this
+        mapping specify the components the M value applies to exactly in order
+        to avoid ambiguity.
+        """
+    ))
+    CONFIG.declare('reduce_bound_constraints', ConfigValue(
+        default=False,
```

Comment:
I put it has False because I thought in some sense the behavior is "unexpected" when it's True--that is it makes a structure different than rote multiple bigm would. But there's no mathematical or performance reason not to let it default to True. It's actually a nicer formulation, and it saves a lot of work if you are calculating M values, especially.